opal/mca/accelerator/accelerator.h

```diff
@@ -0,0 +1,475 @@
+    opal_accelerator_event_t *event);
+
+/**
+ * Copies memory asynchronously from src to dest. Memory of dest and src
+ * may not overlap. Optionally can specify the transfer type to
+ * avoid pointer detection for performance.
+ *
+ * @param[IN] dev_id         Associated device to copy to/from
+ * @param[IN] dest           Destination to copy memory to
+ * @param[IN] src            Source to copy memory from
+ * @param[IN] size           Size of memory to copy
+ * @param[IN] stream         Stream to perform asynchronous copy on
+ * @param[IN] type           Transfer type field for performance
+ *                           Can be set to MCA_ACCELERATOR_TRANSFER_UNSPEC
+ *                           if caller is unsure of the transfer direction.
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_memcpy_async_fn_t)(
+    int dev_id, void *dest, const void *src, size_t size,
+    opal_accelerator_stream_t *stream, opal_accelerator_transfer_type_t type);
+
+/**
+ * Copies memory synchronously from src to dest. Memory of dest and src
+ * may not overlap. Optionally can specify the transfer type to
+ * avoid pointer detection for performance.
+ *
+ * @param[IN] dev_id         Associated device to copy to/from
+ * @param[IN] dest           Destination to copy memory to
+ * @param[IN] src            Source to copy memory from
+ * @param[IN] size           Size of memory to copy
+ * @param[IN] type           Transfer type field for performance
+ *                           Can be set to MCA_ACCELERATOR_TRANSFER_UNSPEC
+ *                           if caller is unsure of the transfer direction.
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_memcpy_fn_t)(
+    int dev_id, void *dest, const void *src, size_t size,
+    opal_accelerator_transfer_type_t type);
+
+/**
+ * Copies a matrix of memory (height rows of width bytes) synchronously
+ * from src to dest. Memory of dest and src may not overlap. Optionally
+ * can specify the transfer type to avoid pointer detection for
+ * performance.
+ *
+ * @param[IN] dev_id         Associated device to copy to/from
+ * @param[IN] dest           Destination to copy memory to
+ * @param[IN] dpitch         Pitch of destination memory
+ * @param[IN] src            Source to copy memory from
+ * @param[IN] spitch         Pitch of source memory
+ * @param[IN] width          Width of matrix transfer (columns in bytes)
+ * @param[IN] height         Height of matrix transfer (rows)
+ * @param[IN] type           Transfer type field for performance
+ *                           Can be set to MCA_ACCELERATOR_TRANSFER_UNSPEC
+ *                           if caller is unsure of the transfer direction.
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_matrix_memcpy_fn_t)(
```

Comment:
which accelerator type need such API?

Reply:
cuda has a 2d memcpy, it's for optimization purposes for something like collectives

Reply:
Is the matrix_memcpy function pointer actually used within OpenMPI anywhere?   This looks like something that is unnecessary and should be removed.  CUDA also has a 3D memcpy, but adding CUDA-specific special things to the function pointers isn't a great idea, unless they are actually used in this PR.   (As mentioned earlier, if desired, they can be added in a future 'v2' of this internal API)

Reply:
It's not currently used. This was a request from @bosilca to improve collectives. It shouldn't be too much of a problem to keep it in since a matrix memcpy can be implemented from regular memcpy afaik

Reply:
Moved matrix memcpy out of this PR.